src/main/java/org/jitsi/videobridge/BitrateController.java

```diff
@@ -0,0 +1,613 @@
+
+                // Route all encodings to the specified bitrate controller.
+                for (int j = 0; j < rtpEncodings.length; j++)
+                {
+                    ssrcToBitrateController.put(
+                        (int) rtpEncodings[j].getPrimarySSRC(), ctrl);
+
+                    if (rtpEncodings[j].getRTXSSRC() != -1)
+                    {
+                        ssrcToBitrateController.put(
+                            (int) rtpEncodings[j].getRTXSSRC(), ctrl);
+                    }
+                }
+            }
+
+            if (ctrl != null)
+            {
+                ctrl.update(targetIdx);
+            }
+
+            if (targetIdx > -1)
+            {
+                newForwardedEndpoints.add(allocations[i].endpointID);
+            }
+        }
+
+        this.forwardedEndpoints = newForwardedEndpoints;
+    }
+
+    /**
+     * Gets the {@link List} of endpoints that are currently being forwarded,
+     * represented by their IDs.
+     *
+     * @return the {@link List} of endpoints that are currently being forwarded,
+     * represented by their IDs.
+     */
+    List<String> getForwardedEndpoints()
+    {
+        return forwardedEndpoints;
+    }
+
+    /**
+     * Checks whether RTP packets from {@code source} should be forwarded
+     * to {@link #dest}.
+     * @param source the channel.
+     * @return {@code true} iff RTP packets from {@code source} should
+     * be forwarded to {@link #dest}.
+     */
+    boolean isForwarded(RtpChannel source)
+    {
+        String endpointID = source.getEndpoint().getID();
+        return forwardedEndpoints.contains(endpointID);
+    }
+
+    /**
+     * Computes the optimal and the target bitrate, limiting the target to be
+     * less than bandwidth estimation specified as an argument.
+     *
+     * @param bweBps the max bandwidth estimation that the target bitrate must
+     * not exceed.
+     *
+     * @return an array of {@link EndpointBitrateAllocation}.
+     */
+    private EndpointBitrateAllocation[] allocate(
+        long bweBps, List<Endpoint> conferenceEndpoints)
+    {
+        // Init.
+        int szConference = conferenceEndpoints.size();
+
+        // subtract 1 for destEndpoint.
+        EndpointBitrateAllocation[] endpointBitrateAllocations
+            = new EndpointBitrateAllocation[szConference - 1];
+
+        int lastN = dest.getLastN();
+        if (lastN < 0)
+        {
+            // If lastN is disable, pretend lastN == szConference.
+            lastN = endpointBitrateAllocations.length;
+        }
+
+        Endpoint destEndpoint = dest.getEndpoint();
+        Set<String> pinnedEndpoints = destEndpoint.getPinnedEndpoints();
+        int szPinned = pinnedEndpoints == null ? 0 : pinnedEndpoints.size();
+        if (szPinned > 0)
+        {
+            // The lastN value should be large enough to accommodate all the
+            // pinned endpoints.
+            lastN = Math.max(lastN, szPinned);
```

Comment:
This changes the pinned semantics. Is it intentional? https://github.com/jitsi/jitsi-videobridge/blob/master/src/main/java/org/jitsi/videobridge/LastNController.java#L494